### PR TITLE
Fixed validation of hostname

### PR DESCRIPTION
**Pull request by mchf**
https://trello.com/c/VrZGjzNP

## Problem ##

Empty hostname was not allowed and erasing /etc/hostname was impossible using yast.

## Solution ##

Fixed the way how yast validates and stores hostname

## Before ##

hostname-old

## After ##

hostname-new

### CODE REVIEW

File: src/lib/y2network/sysconfig/hostname_writer.rb [hunk 1]
```diff
@@ -73,8 +73,15 @@ def update_sysconfig_dhcp(hostname, old_hostname)
       #
       # @param hostname [Y2Network::Hostname] Hostname configuration
       def update_hostname(hostname)
-        Yast::Execute.on_target!("/usr/bin/hostname", hostname.hostname.split(".")[0])
-        Yast::SCR.Write(Yast::Path.new(".target.string"), HOSTNAME_PATH, "#{hostname.hostname}\n")
+        hostname = hostname.static
+        # 1) when user asked for ereasing hostname from /etc/hostname, we keep runtime as it is
```
Comment (teclator) [inline src/lib/y2network/sysconfig/hostname_writer.rb]:
```suggestion
        # 1) when user asked for erasing hostname from /etc/hostname, we keep runtime as it is
```

Comment (teclator) [reply src/lib/y2network/sysconfig/hostname_writer.rb]: In case of erasing and no dhcp, shouldn't it be `localhost`?

Comment (teclator) [reply src/lib/y2network/sysconfig/hostname_writer.rb]: Note: We probably should use hostnamectl in most of the cases, and systemd-firstboot during installation, but it is something we can discuss about.

Comment (mchf) [reply src/lib/y2network/sysconfig/hostname_writer.rb]: well, empty hostname is treated as invalid by hostname utility and empty /etc/hostname is automatically changed to ```localhost``` during the boot.

However, ```hostname ''``` even ```hostname -b --file /etc/hostname``` do not work (expected in first case, but not so expected in the second one according to the man pages). So, runtime remains untouched (hostname cmd still returns previous hostname).

```hostnamectl set-hostname ''``` works in a sense that it deletes /etc/hostname, erases static hostname and sets transient hostname to localhost, so ```hostname``` cmd then correctly shows ```localhost``` as result. However, things like bash promt remains untouched (which is also true when modifying hostname to a nonempty string).

We can put localhost directly into /etc/hostname to mimic the behavior but in my POV it is something different than user wanted to achieve.

So, we can do exactly what hostnamectl does - delete ```/etc/hostname```. However, until we switch to ```hostnamectl ``` we most probably should warn user by a popup that runtime hostname stays untouched until reboot..

**Comment by mchf** [comment]
yep tests are on the way ... once we confirm the solution ;-)

File: src/include/network/services/dns.rb [hunk 2]
```diff
@@ -64,7 +64,7 @@ def initialize_network_services_dns(include_target)
       @widget_descr_dns = {
         "HOSTNAME"        => {
           "widget"            => :textentry,
-          "label"             => Label.HostName,
+          "label"             => "Static H&ostname",
```
Comment (kobliha) [inline src/include/network/services/dns.rb]: _("...")?

Comment (mchf) [reply src/include/network/services/dns.rb]: true but without question mark ;-)

File: test/dns_service_test.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  subject { DummyDnsService.new }
+
+  describe "#ValidateHostname" do
+    it "allows empty hostname" do
+      allow(Yast::UI).to receive(:QueryWidget).and_return("")
+
+      expect(subject.ValidateHostname("", {})).to be true
+    end
+
+    it "allows valid characters in hostname" do
```
Comment (teclator) [inline test/dns_service_test.rb]: But a FQDN is allowed or not?, it is not described by the test, so would also add it. The test should be ok for those who knows which are the valid characters.

File: src/include/network/services/dns.rb [hunk 4]
```diff
@@ -433,15 +429,10 @@ def HandleResolverData(key, _event)
     # @param _event [Hash] the event being handled
     # @return whether valid
```
Comment (teclator) [inline src/include/network/services/dns.rb]:
```suggestion
    # @return [Boolean] whether the current static hostname is valid or not
```

File: test/dns_service_test.rb (hunk 3 above)
Comment (teclator) [reply test/dns_service_test.rb]: BTW, the valid_chars check is done by the widget, and looks a test more for it than for the validate method

File: src/lib/y2network/sysconfig/hostname_writer.rb [hunk 5]
```diff
@@ -74,7 +74,7 @@ def update_sysconfig_dhcp(hostname, old_hostname)
       # @param hostname [Y2Network::Hostname] Hostname configuration
       def update_hostname(hostname)
```
Comment (teclator) [inline src/lib/y2network/sysconfig/hostname_writer.rb]: What about adding a unit test for this writer?

File: test/dns_service_test.rb (hunk 3 above)
Comment (mchf) [reply test/dns_service_test.rb]: true ... updated with test for FQDN

File: src/lib/y2network/sysconfig/hostname_writer.rb (hunk 5 above)
Comment (teclator) [reply src/lib/y2network/sysconfig/hostname_writer.rb]: It seems that the `SENDMAIL` ... constant is not used, so probably copied from the `dns_writer`

https://github.com/yast/yast-network/blob/e7d6c742f4617ba4dd0cbd5cdbeba0ece4053195/src/lib/y2network/sysconfig/hostname_writer.rb#L45

**Comment by teclator** [comment]
np: a screenshot would be nice to have and maybe a TODO section with action items to be done in the future, like for example show the current dhcp hostname

File: test/y2network/hostname_writer_test.rb [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+require_relative "../test_helper"
+require "y2network/sysconfig/hostname_writer"
+require "y2network/hostname"
+
+describe Y2Network::Sysconfig::HostnameWriter do
+  subject { Y2Network::Sysconfig::HostnameWriter.new }
+
+  describe ".update_hostname" do
```
Comment (teclator) [inline test/y2network/hostname_writer_test.rb]: Update hostname is private, and you are really testing the write call that is ok.

```suggestion
  describe "#write" do
```

File: test/y2network/hostname_writer_test.rb [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+    let(:old_hostname_container) do
+      instance_double(
+        Y2Network::Hostname,
+        static:         "old#{hostname}",
+        dhcp_hostname:  false,
+        save_hostname?: true
+      )
+    end
+
```
Comment (teclator) [inline test/y2network/hostname_writer_test.rb]: Miss when no change context

File: test/y2network/hostname_writer_test.rb (hunk 6 above)
Comment (teclator) [inline test/y2network/hostname_writer_test.rb]: please mock the call to the syconfig dhcp update.

File: test/y2network/hostname_writer_test.rb [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+          .with(anything, anything, /#{hostname}/)
+
+        subject.write(hostname_container, old_hostname_container)
+      end
+    end
+
+    context "when no change in hostname" do
+      let(:hostname) { "hostname" }
+
+      it "do not try to update anything" do
```
Comment (teclator) [inline test/y2network/hostname_writer_test.rb]:
```suggestion
      it "does not try to update anything" do
```

File: test/y2network/hostname_writer_test.rb [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        expect(Yast::Execute)
+          .not_to receive(:on_target!)
+        expect(Yast::SCR)
+          .not_to receive(:Write)
+          .with(anything, anything, /#{hostname}/)
```
Comment (teclator) [inline test/y2network/hostname_writer_test.rb]: you can omit the .with

Comment (teclator) [inline test/y2network/hostname_writer_test.rb]:
```suggestion
```

File: test/y2network/hostname_writer_test.rb [hunk 10]
```diff
@@ -0,0 +1,92 @@
+# Copyright (c) [2019] SUSE LLC
```
Comment (teclator) [inline test/y2network/hostname_writer_test.rb]:
```suggestion
# Copyright (c) [2020] SUSE LLC
```

File: test/dns_service_test.rb [hunk 11]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env rspec
+
+# Copyright (c) [2019] SUSE LLC
```
Comment (teclator) [inline test/dns_service_test.rb]:
```suggestion
# Copyright (c) [2020] SUSE LLC
```

**Pull request merged**